src/__tests__/retry-queue.js

```diff
@@ -0,0 +1,195 @@
+            options: defaultRequestOptions,
+            requestId: 4,
+        })
+    }
+
+    beforeEach(() => {
+        jest.useFakeTimers()
+        jest.spyOn(given.queue, 'getTime').mockReturnValue(EPOCH)
+    })
+
+    it('processes retry requests', () => {
+        enqueueRequests()
+
+        expect(given.queue._counterToQueueMap[1].length).toEqual(4)
+        expect(given.queue._requestRetriesMap).toEqual({
+            1: 0,
+            2: 0,
+            3: 0,
+            4: 0,
+        })
+
+        given.queue.poll()
+
+        expect(window.XMLHttpRequest).toHaveBeenCalledTimes(0)
+
+        jest.runOnlyPendingTimers()
+
+        expect(window.XMLHttpRequest).toHaveBeenCalledTimes(4)
+
+        expect(given.queue._counterToQueueMap[1]).toEqual(undefined)
+
+        enqueueRequests()
+
+        expect(given.queue._counterToQueueMap[2].length).toEqual(4)
+        expect(given.queue._requestRetriesMap).toEqual({
+            1: 1,
+            2: 1,
+            3: 1,
+            4: 1,
+        })
+
+        expect(given.queue._counterToQueueMap[2]).toEqual([
+            {
+                url: '/e',
+                data: { event: 'foo', timestamp: EPOCH - 3000 },
+                options: defaultRequestOptions,
+                requestId: 1,
+            },
+            {
+                url: '/e',
+                data: { event: 'bar', timestamp: EPOCH - 2000 },
+                options: defaultRequestOptions,
+                requestId: 2,
+            },
+            {
+                url: '/e',
+                data: { event: 'baz', timestamp: EPOCH - 1000 },
+                options: defaultRequestOptions,
+                requestId: 3,
+            },
+            {
+                url: '/e',
+                data: { event: 'fizz', timestamp: EPOCH },
+                options: defaultRequestOptions,
+                requestId: 4,
+            },
+        ])
+    })
+
+    it('tries to send requests via beacon on unload', () => {
+        enqueueRequests()
+
+        given.queue.poll()
+        given.queue.unload()
+
+        expect(given.queue._event_queue.length).toEqual(0)
+        expect(window.navigator.sendBeacon).toHaveBeenCalledTimes(4)
+    })
+
+    it('enqueues requests when offline and flushes immediately when online again', () => {
+        enqueueRequests()
+
+        given.queue._areWeOnline = false
+        expect(given.queue._areWeOnline).toEqual(false)
+
+        given.queue.poll()
+        jest.runOnlyPendingTimers()
+
+        expect(window.XMLHttpRequest).toHaveBeenCalledTimes(4)
+        expect(given.queue._offlineBacklog.length).toEqual(4)
+
+        given.queue._handleWeAreNowOnline()
+
+        expect(given.queue._areWeOnline).toEqual(true)
+        expect(given.queue._offlineBacklog.length).toEqual(0)
+
+        expect(window.XMLHttpRequest).toHaveBeenCalledTimes(8)
+    })
+
+    it('retries using an exponential backoff mechanism', () => {
```

Comment:
Could test this with a loop too and go all the way but thought it would be a bit convoluted